subprojects/files/src/main/java/org/gradle/api/internal/file/collections/BuildDependenciesOnlyFileCollectionResolveContext.java

```diff
@@ -61,10 +59,11 @@ public FileCollectionResolveContext add(Object element) {
         } else if (element instanceof TaskOutputs) {
             TaskOutputs outputs = (TaskOutputs) element;
             taskContext.add(outputs.getFiles());
-        } else if (element instanceof RegularFileProperty || element instanceof DirectoryProperty) {
-            taskContext.add(element);
-        } else if (element instanceof TaskProvider) {
-            taskContext.add(((TaskProvider) element).get());
+        } else if (element instanceof ProviderInternal) {
+            ProviderInternal<?> provider = (ProviderInternal<?>) element;
+            if (!provider.maybeVisitBuildDependencies(taskContext)) {
+                taskContext.maybeAdd(((ProviderInternal<?>) element).get());
```

Comment:
You already cast `element` to `provider`, you don't need to cast it again here do you?